sssd_test_framework/utils/dbus_types.py

```diff
@@ -0,0 +1,675 @@
+
+    _type_prefix = "variant"
+
+    def __init__(self, model_element: DBUSType | None = None):
+        """
+        An object of the expected type can be passed as parameter.
+        This object will be used as a model.
+
+        If the object is not passed, we will try to guess it. This behavior is
+        required because the variant type doesn't include the subtype in the
+        signature.
+        """
+        self._element: DBUSType | None = None if model_element is None else model_element.clone()
+        """The element stored by the variant"""
+
+    @property
+    def value(self):
+        if self._element is None:
+            raise ValueError("No value set")
+        return self._element.value
+
+    @value.setter
+    def value(self, val):
+        if self._element is None:
+            raise ValueError("No subtype set for Variant")
+        self._element.value = val
+
+    def __str__(self) -> str:
+        return str(self._element)
+
+    def clone(self):
+        return self.__class__(self._element)
+
+    # Parsing:  variant <type> <value>
+    def parse(self, result: DBUSResult) -> None:
+        res = re.match(r"^ *variant +", result.text)
+        if res is None:
+            raise TypeError("Not a variant")
+
+        old_text = result.text
+        result.text = result.text.removeprefix(res.group(0))
+
+        if self._element is None:
+            self._element = DBUSType._guess_next_type(result)
+
+        try:
+            self._element.parse(result)
+        except Exception as e:
+            result.text = old_text
+            raise e
+
+    # Param:  variant:<type>:<value>
+    def param(self) -> str:
+        if self._element is None:
+            raise ValueError("No value set")
+        return f"{self._type_prefix}:{self._element.param()}"
+
+
+class DBUSTypeDict(DBUSTypeContainer):
+
+    _type_prefix = "dict"
+
+    def __init__(self, model_key: DBUSType | None = None, model_value: DBUSType | None = None):
+        """
+        Build a dictionary containing elements of a single type and keys of
+        another type.
+
+        An object of each type can be passed as parameter. This object will be
+        cloned to create the elements when values are assigned.
+
+        If the objects are not passed, we will try to guess them. This behavior
+        is required because the variant type doesn't include the subtype in
+        the signature.
+        """
+        if (model_key is None or model_value is None) and model_key != model_value:
+            raise ValueError("Both arguments or none must be None")
+
+        self._model_key = model_key
+        self._model_value = model_value
+        self._str: str | None = None
+
+    @property
+    def value(self) -> dict:
+        return dict(self._value)
+
+    @value.setter
+    def value(self, val: dict):
+        if self._model_key is None or self._model_value is None:
+            raise ValueError("No subtype set")
+
+        self._value = dict(val)
```

Comment:
It's done on purpose. On one side this will check the passed value is a `dict` as requested, and on the other side it will create a new `dict` so that we don't kept the caller's object which they can modify after it was assigned.

Reply:
Ok.